Added more detail to the section on removing images/containers (#11)
dme26/docker-introduction: _episodes/03-creating-containers.md

~~~
Unable to find image 'hello-world:latest' locally
latest: Pulling from library/hello-world

> 1b930d010525: Pull complete

Reviewer: I'm completely happy to merge this sort of change, for convenience, but it just looks like whitespace. Is this possibly an editing artefact?

Author: ```suggestion
1b930d010525: Pull complete 
```

Author: I think this is an artifact from editing.  I can switch it back but it will make another commit.  However, when you merge this I think you should get the option to squash all of my commits into one so it may not matter.  Commit my suggestion above if that sounds good to you.

~~~
{: .output}

### Removing images

> If you need to reclaim space, you will need to remove image files.

Reviewer: How about "reclaim disk space" ?
```suggestion
If you need to reclaim disk space, you can remove image files.
```

> The images and their corresponding containers can start to take up a lot of disk space of you don't clean them up occasionally.

Reviewer: Just fixing a small typo.
```suggestion
The images and their corresponding containers can start to take up a lot of disk space if you don't clean them up occasionally.
```

~~~
$ docker image ls
~~~
{: .language-bash}
~~~
REPOSITORY             TAG                 IMAGE ID            CREATED             SIZE
ubuntu                 trusty              dc4491992653        12 months ago       222MB
hello-world            latest              fce289e99eb9        4 months ago        1.84kB
~~~
{: .output}

We can try to remove remove the `hello-world` image with a `docker image rm` command that includes the repository name, like so but...:
~~~
$ docker image rm hello-world
~~~
{: .language-bash}
~~~
Error response from daemon: conflict: unable to remove repository reference "hello-world" (must force) - container e7d3b76b00f4 is using its referenced image fce289e99eb9
~~~
{: .output}
We get this error because there are containers created that depend on this image.

### What containers are running?
As indicated by the error above there is still an existing container from this image.
We can list the running containers by typing.
~~~
$ docker container ls
~~~
{: .language-bash}
~~~
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
~~~
{: .output}

> The first version of the command is ortholognal to what we know from the unix shell but we might want to use the shorter version. For refernece, "ps" stands for “Process Status”.

Reviewer: I typically go with leading caps for Unix; also tweaking two typos.
```suggestion
The first version of the command is orthogonal to what we know from the Unix shell but we might want to use the shorter version. For reference, "ps" stands for “Process Status”.
```

> We should also notice that this command didn't return any containers because our containers all exited and stopped running after they ran.

Reviewer: Avoiding reuse of "ran". Maybe this is slightly clearer?
```suggestion
We should also notice that this command didn't return any containers because our containers all exited and thus stopped running after they completed their work.
```

The first version of the command is ortholognal to what we know from the unix shell but we might want to use the shorter version. For refernece, "ps" stands for “Process Status”.
We should also notice that this command didn't return any containers because our containers all exited and stopped running after they ran.

### What containers have run recently?
There is also a way to list running containers, and those that have completed recently, which is to add the `--all`/`-a` flag to the `docker container ls`/`docker ps` command as shown below.
~~~
$ docker container ls --all
~~~
{: .language-bash}
~~~
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                     PORTS               NAMES
9c698655416a        hello-world         "/hello"            2 minutes ago       Exited (0) 2 minutes ago                       zen_dubinsky
6dd822cf6ca9        hello-world         "/hello"            3 minutes ago       Exited (0) 3 minutes ago                       eager_engelbart
~~~
{: .output}

> ### Conclusion

Reviewer: Do you want to remove the conclusion? I'm happy to do so, but am just checking this was intended.

Author: Nope, that was an accident. Thanks for catching!

Author: ```suggestion
### Conclusion
```

Author: Same issue with extra commits to switch it back.  Sorry! commit suggestion below and I recommend you squash all these commits into one when you merge this PR.

Author: 😦 It looks like it won't let me suggest a change here to add that back.  I'll pull the updates to this PR branch and then fix it and push again.

Author: It added it back below but should be right now.

~~~
$ docker container rm 9c698655416a
~~~
{: .language-bash}
~~~
9c698655416a
~~~
{: .output}

If you want to remove all exited containers at once you can use the `docker containers prune` command.
**Be careful** with this command.
If you have containers you may want to reconnect to, you should not use this command.
It will ask you if to confirm you want to remove these containers, see output below.
If successfull it will print the full `CONTAINER ID` back to you.
~~~
$ docker container prune
~~~
{: .language-bash}
~~~
WARNING! This will remove all stopped containers.
Are you sure you want to continue? [y/N] y
Deleted Containers:
9c698655416a848278d16bb1352b97e72b7ea85884bff8f106877afe0210acfc
6dd822cf6ca92f3040eaecbd26ad2af63595f30bb7e7a20eacf4554f6ccc9b2b
~~~
{: .output}

### Removing images, for real this time

~~~
$ docker image rm hello-world
~~~
{: .language-bash}
~~~
Untagged: hello-world:latest
Untagged: hello-world@sha256:5f179596a7335398b805f036f7e8561b6f0e32cd30a32f5e19d17a3cda6cc33d
Deleted: sha256:fce289e99eb9bca977dae136fbe2a82b6b7d4c372474c9235adc1741675f587e
Deleted: sha256:af0b15c8625bb1938f1d7b17081031f649fd14e6b233688eea3c5483994a66a3
~~~
{: .output}

> The reason that there is a few lines output, is that a given image may be created by merging multiple underlying layers.

Reviewer: Two potential grammar tweaks.
```suggestion
The reason that there are a few lines of output, is that a given image may have been formed by merging multiple underlying layers.
```
